streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -395,8 +395,10 @@ private void createNewTasks(final Map<TaskId, Set<TopicPartition>> activeTasksTo
             tasks.addActiveTasks(newActiveTasks);
             tasks.addStandbyTasks(newStandbyTask);
         } else {
-            tasks.addPendingTaskToInit(newActiveTasks);
-            tasks.addPendingTaskToInit(newStandbyTask);
+            final Map<TaskId, RuntimeException> taskInitExceptions = new LinkedHashMap<>();
+            Stream.concat(newActiveTasks.stream(), newStandbyTask.stream())
+                    .forEach(t -> addTaskToStateUpdater(t, taskInitExceptions));
```

Comment:
When I started working on the state updater I had the same thought, but then I learned that task initialization might take some time due to the need of opening RocksDB instances. Having such time consuming operations on the rebalance path is not ideal. I do not remember exactly the reason, but I think with cooperative rebalancing the client might miss the immediate next rebalance which would lead again to more rebalances.
@ableegoldman Please keep me honest here.

Reply:
If that is the case, especially if it takes a long time to init rocksdb, maybe it would be the be easier to initialize the state in the state updater instead?

Reply:
Yeah, we considered also to initialise the tasks inside the state updater, but we decided to do it outside so that we do not need to change the state of a task inside the state updater. It seemed cleaner to us to keep lifecycle management away from the state updater. The state updater should only update the states.
What are the advantages of moving the initialisation into the state updater? I see that we would not need to explicitly handle exceptions during intialisation of tasks since that would happen automatically within the state updater. Is there something else?

Reply:
If we initialize inside the state updater (lazily using `initializeIfNeeded`, before restoration), we should be able to avoid any extra handling for these tasks. In `handleAssignment`, we can directly add the uninitialized tasks to the state updater.

The advantages would be
- No `pendingTaskToInit` required, so some simplification.
- We don't crash when two rebalances happen in one round of polling (why I opened this PR, see description)
- After a rebalance, all tasks that remained with this instance can immediately continue processing, and we do not have to wait for new tasks to initialize their rocksdb (this is more the cherry on top)

The last option that I see to fix the `IllegalStateException` (besides moving initialization to the poll-phase or to the state updater) is that we can take care to not recreate tasks that are already in `pendingTaskToInit`, but it could get complicated. It can happen that the input partitions change, in which case we need to move the already created task but not initialized task to `pendingTaskToUpdateInputPartitions` - but at the same time remember that the task isn't initialized yet ... not sure exactly how to handle these things correctly, but it looks like it will just add more corner cases to think about.

Reply:
I think I am convinced. Let's give initializing the tasks inside the state updater a try.